File: vscode-extension/src/logic.js
// @ts-check
"use strict";

/**
 * Pure decoration rules shared by the extension and unit tests.
 * Keep this free of the `vscode` module so CI can run it with plain Node.
 */

const IMPURE_CODES = new Set(["PURE001", "PURE002"]);
const PURE_CODES = new Set(["PURE003"]);

/** Badge text shown in the editor (end-of-line `after` content). */
const BADGE_IMPURE = "impure";
const BADGE_PURE = "pure";

/**
 * @param {unknown} code
 * @returns {string}
 */
function diagnosticCode(code) {
  if (typeof code === "object" && code !== null) {
    const value = /** @type {{ value?: unknown }} */ (code).value;
    return String(value ?? "");
  }
  return String(code ?? "");
}

/**
 * @param {{ code?: unknown, source?: unknown }} d
 */
function isPureAnalyzerDiagnostic(d) {
  const code = diagnosticCode(d.code);
  const source = String(d.source ?? "");
  return (
    IMPURE_CODES.has(code) ||
    PURE_CODES.has(code) ||
    source.includes("Pure analyzer") ||
    source.includes("FSharp.PureAnalyzer")
  );
}

/**
 * Map definition diagnostics to per-line badges.
 * Only PURE002 / PURE003 produce badges (PURE001 is call-site only).
 * Impure wins over pure on the same line.
 *
 * @param {Array<{ code?: unknown, source?: unknown, range?: { start?: { line?: number } } }>} diagnostics
 * @returns {Map<number, { badge: "impure" | "pure", code: string }>}
 */
function badgesByLine(diagnostics) {
  /** @type {Map<number, { impure?: string, pure?: string }>} */
  const byLine = new Map();

  for (const d of diagnostics) {
    if (!isPureAnalyzerDiagnostic(d)) continue;
    const code = diagnosticCode(d.code);
    const line = d.range?.start?.line;
    if (typeof line !== "number") continue;

    let entry = byLine.get(line);
    if (!entry) {
      entry = {};
      byLine.set(line, entry);
    }
    if (code === "PURE002") {
      entry.impure = code;
    } else if (code === "PURE003") {
      entry.pure = code;
    }
  }

  /** @type {Map<number, { badge: "impure" | "pure", code: string }>} */
  const result = new Map();
  for (const [line, entry] of byLine) {
    if (entry.impure) {
      result.set(line, { badge: BADGE_IMPURE, code: entry.impure });
    } else if (entry.pure) {
      result.set(line, { badge: BADGE_PURE, code: entry.pure });
    }
  }
  return result;
}

/**
 * @param {string} code  e.g. PURE002
 * @returns {"impure" | "pure" | null}
 */
function badgeForDefinitionCode(code) {
  if (code === "PURE002") return BADGE_IMPURE;
  if (code === "PURE003") return BADGE_PURE;
  return null;
}

module.exports = {
  IMPURE_CODES,
  PURE_CODES,
  BADGE_IMPURE,
  BADGE_PURE,
  diagnosticCode,
  isPureAnalyzerDiagnostic,
  badgesByLine,
  badgeForDefinitionCode,
};

File: src/Fspure.DecorationLogic/Logic.fs
namespace Fspure.DecorationLogic

open System
open System.Collections.Generic

#if FABLE_COMPILER
open Fable.Core.JsInterop
#endif

/// Pure decoration rules shared by the VS Code extension (via Fable) and .NET tests.
module Logic =

    let impureCodes = set [ "PURE001"; "PURE002" ]
    let pureCodes = set [ "PURE003" ]

    let badgeImpure = "impure"
    let badgePure = "pure"

    /// Normalize diagnostic code from string or { value: string } shape.
    let diagnosticCode (code: obj) : string =
#if FABLE_COMPILER
        if isNull code then
            ""
        elif jsTypeof code = "string" then
            unbox<string> code
        else
            let v: obj = code?value

            if isNull v then "" else string v
#else
        match code with
        | null -> ""
        | :? string as s -> s
        | o ->
            let t = o.GetType()
            let p = t.GetProperty("value")

            if isNull p then string o
            else
                match p.GetValue(o) with
                | null -> ""
                | v -> string v
#endif

    type DiagnosticLike =
        {
            Code: string
            Source: string
            Line: int option
        }

    let isPureAnalyzerDiagnostic (d: DiagnosticLike) : bool =
        let code = d.Code
        let source = d.Source

        impureCodes.Contains code
        || pureCodes.Contains code
        || source.IndexOf("Pure analyzer", StringComparison.Ordinal) >= 0
        || source.IndexOf("FSharp.PureAnalyzer", StringComparison.Ordinal) >= 0

    type BadgeEntry = { Badge: string; Code: string }

    /// Map definition diagnostics to per-line badges.
    /// Only PURE002 / PURE003 produce badges (PURE001 is call-site only).
    /// Impure wins over pure on the same line.
    let badgesByLine (diagnostics: DiagnosticLike seq) : Map<int, BadgeEntry> =
        let byLine = Dictionary<int, string option * string option>()

        for d in diagnostics do
            if isPureAnalyzerDiagnostic d then
                match d.Line with
                | None -> ()
                | Some line ->
                    let code = d.Code

                    let impure', pure' =
                        match byLine.TryGetValue line with
                        | true, v -> v
                        | _ -> None, None

                    let next =
                        if code = "PURE002" then Some code, pure'
                        elif code = "PURE003" then impure', Some code
                        else impure', pure'

                    byLine[line] <- next

        [
            for KeyValue(line, (impure', pure')) in byLine do
                match impure' with
                | Some c -> yield line, { Badge = badgeImpure; Code = c }
                | None ->
                    match pure' with
                    | Some c -> yield line, { Badge = badgePure; Code = c }
                    | None -> ()
        ]
        |> Map.ofList

    let badgeForDefinitionCode (code: string) : string option =
        if code = "PURE002" then Some badgeImpure
        elif code = "PURE003" then Some badgePure
        else None
